test/test.js

```diff
@@ -235,6 +235,46 @@ describe('Page', function() {
     }));
   });
 
+  describe('Browser.Events.disconnected', function() {
+    it('should emitted when browser gets closed', SX(async function() {
+      const browser = await puppeteer.launch(defaultBrowserOptions);
+      let isDisconnected = false;
+      browser.on('disconnected', dc => {
+        expect(dc).toBe('connection disconnected');
+        isDisconnected = true;
+      });
+      await browser.close();
+      expect(isDisconnected).toBe(true);
+    }));
+    it('should emitted when browser.disconnect() called', SX(async function() {
+      const originalBrowser = await puppeteer.launch(defaultBrowserOptions);
+      const newBrowser = await puppeteer.connect({
+        browserWSEndpoint: originalBrowser.wsEndpoint()
+      });
+      let isDisconnected = false;
+      newBrowser.on('disconnected', dc => {
+        isDisconnected = true;
+      });
+      newBrowser.disconnect();
+      expect(isDisconnected).toBe(true);
+    }));
+    it('should be emitted when underlying websocket gets closed', SX(async function() {
+      const browser = await puppeteer.launch(defaultBrowserOptions);
+      let resolver = null;
+      const promise = new Promise((res, rej) => {
+        resolver = res;
+      });
+      let isDisconnected = false;
+      browser.on('disconnected', dc => {
+        isDisconnected = true;
+        resolver();
+      });
+      browser._connection._ws.close(3006, 'intentionally kill it');
```

Comment:
we should not reach to private variables in tests.
Instead, this test could be written like the previous one, but instead of disconnecting
the new browser, closing original one:

```js
it('should be emitted when underlying websocket gets closed', SX(async function() {
  const browser = await puppeteer.launch(defaultBrowserOptions);
  const newBrowser = await puppeteer.connect({
    browserWSEndpoint: originalBrowser.wsEndpoint()
  });
  let isDisconnected = false;
  newBrowser.once('disconnected', () => isDisconnected = true);
  await originalBrowser.close();
  expect(isDisconnected).toBe(true);
}));

```

Reply:
Sorry i didn't know that, i just try to get closer to 
> should be emitted when underlying websocket gets closed

Most of my time yesterday was spend here, i wish know it earlier. Next time i should be more clever on testing.

Reply:
@brutalcrozt feel free to ask questions next time if you're getting stuck.